Keas.Mvc/ClientApp/components/Keys/EditKey.tsx

```diff
@@ -1,151 +1,123 @@
-  public static contextType = Context;
-  public context!: React.ContextType<typeof Context>;
-
-  private _formRef: React.RefObject<HTMLFormElement>;
-
-  constructor(props) {
-    super(props);
-
-    this._formRef = React.createRef();
-
-    this.state = {
-      error: {
-        message: '',
-        path: ''
-      },
-      key: this.props.selectedKey,
-      submitting: false,
-      validState: false
-    };
-  }
-
-  public render() {
-    if (!this.state.key) {
-      return null;
-    }
+const EditKey = (props: IProps) => {
+  const [error, setError] = useState<IValidationError>({
+    message: '',
+    path: ''
+  });
+  const [key, setKey] = useState<IKey>(props.selectedKey);
+  const [submitting, setSubmitting] = useState<boolean>(false);
+  const [validState, setValidState] = useState<boolean>(false);
+  const formRef: React.RefObject<HTMLFormElement> = useRef(null);
+
+  const validateState = () => {
+    const checkIfKeyCodeIsValid = props.checkIfKeyCodeIsValid;
+    const error = yupAssetValidation(keySchema, key, {
+      context: { checkIfKeyCodeIsValid }
+    });
+    setError(error);
+    setValidState(error.message === '');
+  };
 
-    const { searchableTags } = this.props;
+  useEffect(() => {
+    validateState();
+  }, [key]);
 
-    return (
-      <Modal
-        isOpen={this.props.modal}
-        toggle={this._confirmClose}
-        size='lg'
-        className='keys-color'
-      >
-        <div className='modal-header row justify-content-between'>
-          <h2>Edit Key</h2>
-          <Button color='link' onClick={this._closeModal}>
-            <i className='fas fa-times fa-lg' />
-          </Button>
-        </div>
-        <ModalBody>
-          <div className='container-fluid'>
-            <form ref={this._formRef}>
-              <KeyEditValues
-                selectedKey={this.state.key}
-                changeProperty={this._changeProperty}
-                disableEditing={false}
-                searchableTags={searchableTags}
-                error={this.state.error}
-              />
-            </form>
-          </div>
-        </ModalBody>
-        <ModalFooter>
-          <Button
-            color='primary'
-            onClick={this._editSelected}
-            disabled={!this.state.validState || this.state.submitting}
-          >
-            Go!{' '}
-            {this.state.submitting && (
-              <i className='fas fa-circle-notch fa-spin' />
-            )}
-          </Button>{' '}
-        </ModalFooter>
-      </Modal>
-    );
+  if (!key) {
+    return null;
   }
 
-  private _changeProperty = (property: string, value: string) => {
-    this.setState(
-      prevState => ({
-        key: {
-          ...prevState.key,
-          [property]: value
-        }
-      }),
-      this._validateState
-    );
+  const { searchableTags } = props;
+  const changeProperty = (property: string, value: string) => {
+    key[property] = value;
+    setKey(key);
```

Comment:
@wongband this is not actually the same thing, and I'm not sure if it'll create error or not but you should verify your useEffect is being called if you go this way.  Ex: `key[property] = value; setKey(key);` like you did means that `key` is the same object reference the whole time.  Where as `setKey({...key, [property]: value})` creates a whole new object to put into state.  It used to be that you always wanted to replace your object when modifying state, or else the change detection in setState() wouldn't always work properly, especially for nested objects.... Anyway, mainly I mention this so you can be extra sure than when replacing a fully-new object with a simply modified object that it doesn't change behavior and that state is actually updated and related useEffect(()=>{}, [key]) and other hooks still work.

Reply:
I just checked with:
```
key[property] = value;
setKey(key);
```
and
```
    setKey({...key, [property]: value})
```
Both methods calls `useEffect`. I'm assuming that we should go with the second way since that is the standard way to do it?

Reply:
@wongband yes I think so too, especially when we get into nested objects I think it'll be a bigger deal.  I'll read more on the topic (of immutable w/ hooks) but I agree I think standardizing on the second way is a good idea

Reply:
Okay sounds good. I'll change it to that right now